pkg/kubelet/kubelet.go

```diff
@@ -884,6 +884,9 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 		volumepathhandler.NewBlockVolumePathHandler())
 
 	klet.backOff = flowcontrol.NewBackOff(containerBackOffPeriod, MaxContainerBackOff)
+	klet.backOff.HasExpiredFunc = func(eventTime time.Time, lastUpdate time.Time, maxDuration time.Duration) bool {
+		return eventTime.Sub(lastUpdate) > 600*time.Second
```

Comment:
will there be any side effects if maxDuration is more than `600*time.Second`? Will we clear up the backoff period for a filed pod in this case?

It cannot happen today, but with the configurability we can forget to add this check

Reply:
If maxDuration is more than 600s (which, in the current implementation, cannot happen because it will be validated to maximum 300, but I know you have an opinion about that in the other PR and it could at some point be made unbounded), then the side effect is that those pods that have run successfully for 600s will still have their backoff counter cleared the next time they exit. This still tracks with my design goal as I'm purposefully stabilizing the expiration to 600s regardless of the maxDuration set by the user.  I also do not see this as a problem in general as the risk the backoff counter is trying to cap on is quickly crashing pods, not any pods that crashing -- at their fastest -- every 601 seconds

Reply:
hasExpired won't return true just because the user provided maximum timeout is long and there's a long wall time wait, because it's not comparing if the elapsed time since we last backed off is 600s -- it's comparing if the length of time we last saw the pod run is under 600s. It compares how long its been since the latest exited container of the pod (eventTime) and the last time we ticked a backoff (lastUpdate); the bigger the difference, the longer the pod ran since the last time we moved the backoff counter up. That value will stay stable even if we are in a long backoff because both eventTime and lastUpdate do not change until the next backoff is ticked.

That being said this did make me see another potential bug
/hold

There _is_ also a backoff GC [here](https://github.com/kubernetes/kubernetes/blob/67a4d209708b75a876440df4737104cb30ba5e20/staging/src/k8s.io/client-go/util/flowcontrol/backoff.go#L144) I didn't notice that uses only the internal lastUpdate time AND it's hard coded to 2x max duration. For >300s maxDurations, that's actually fine, because GC will happen well over 600s which is what I care about keeping stable, but for the intended low maxDurations, they might reset "early" because they occasionally would get GCd without checking their special HasExpiredFunc properties. So I definitely need to introduce something that can inject a minimum of 600s [here](https://github.com/kubernetes/kubernetes/blob/67a4d209708b75a876440df4737104cb30ba5e20/staging/src/k8s.io/client-go/util/flowcontrol/backoff.go#L149).